File: Cpp/2336_smallest_number_in_infinite_set.cpp
// https://leetcode.cn/problems/smallest-number-in-infinite-set/

class SmallestInfiniteSet {
public:
    SmallestInfiniteSet() {
        min_val_ = 1;
    }
    
    int popSmallest() {
        removed_.insert(min_val_);
        int min_value = min_val_;
        while (removed_.find(min_val_) != removed_.end()) {
            min_val_ += 1;
        }
        return min_value;
    }
    
    void addBack(int num) {
        if (removed_.find(num) != removed_.end()) {
            removed_.erase(num);
        }
        min_val_ = min(min_val_, num);
    }

    unordered_set<int> removed_;
    int min_val_;
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: Python3/2336_smallest_number_in_infinite_set.py
# https://leetcode.cn/problems/smallest-number-in-infinite-set/

class SmallestInfiniteSet:

    def __init__(self):
        self.removed = set()
        self.min_val = 1

    def popSmallest(self) -> int:
        self.removed.add(self.min_val)
        min_val = self.min_val
        while self.min_val in self.removed:
            self.min_val += 1
        return min_val

    def addBack(self, num: int) -> None:
        if num in self.removed:
            self.removed.remove(num)
        self.min_val = min(self.min_val, num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)